[US-4039] Complete refactoring of feeds functionality

PetrHeinz opened the pull request:
I had some issues with how to write automatic tests for the feeds. I ended up fighting with mocking in PHPUnit and I'm not really satisfied with the result in [GoogleFeedItemTest](https://github.com/shopsys/shopsys/blob/d7520e696c455bb2475848d5b60d88c0513a1cfd/packages/product-feed-google/tests/Unit/GoogleFeedItemTest.php). I was thinking about directly creating the instances of Product instead of using mocks. I will appreciate tips on this.

I've briefly measured the performance improvements with 50k products and it seems that the suggested changes improved the speed of Google feed product export by the factor of ten, which is nice. Also, it was designed with extensibility in mind and it should be now possible to change the export (eg. add a new exported tag) from the project's repository. However, I didn't test it.

I did consult it with @vitek-rostislav during development to avoid dead-ends.

| Q             | A
| ------------- | ---
|Description, reason for the PR| Feeds were slow to generate and inflexible.
|New feature| No <!-- Do not forget to update CHANGELOG.md and possibly docs/ -->
|BC breaks| Yes <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| #208 
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

Miroslav-Stopka commented:
Hi Peter,

thank you for this PULL REQUEST, Im on it :)

Miroslav-Stopka commented on composer.json:
```diff
@@ -15,10 +15,7 @@
       "Shopsys\\MigrationBundle\\": "packages/migrations/src/",
       "Shopsys\\Plugin\\": "packages/plugin-interface/src/",
       "Shopsys\\ProductFeed\\": "packages/product-feed-interface/src/",
-      "Shopsys\\ProductFeed\\ZboziBundle\\": "packages/product-feed-zbozi/src/",
```
First, Im not sure why this was neccessary. Removal of all feeds? Im propably not comfortable with the solution - remove all and create new. This should be the last option if everything fails. Not the first one. With this approach, everyone can remove all while refactoring.

PetrHeinz replied:
No, no, no, you've got this wrong - it's only for the time being - I will drop this commit and refactor the feeds. I only didn't want to deal with them during the work on a single feed.

After the review I will drop the commit and redo the rest of the feeds just like I did with the Google feed.

Miroslav-Stopka replied:
Thank you for clearing up

Miroslav-Stopka commented on packages/product-feed-google/tests/Unit/GoogleFeedTest.php:
```diff
@@ -1,308 +0,0 @@
-use Twig_Environment;
-use Twig_Loader_Filesystem;
-
-class GoogleFeedTest extends TestCase
```
Did you create new test as a replacement for this one? Plus it would be nice to have some test for a real generation with some demo data from db. I think that last changes in multidomains affected a generation of feed too.

Miroslav-Stopka replied:
I agree with the idea of a new test - feeds are exportable on all domains - in smoke testing layer :) . Plus I propose to create at least some test (maybe integration test) using the FeedItem and FeedRenderer with control, whether a specific xml string is returned based on specific product data.

PetrHeinz replied:
I've added the smoke test in 16adeff. I couldn't create the integration test in the feed's package as it would create a dependency on project's repository - there is no kernel in shopsys/framework so there is no DIC and I cannot fetch the services I need...

I could have created the integration test in shopsys/project-base, but then there would be a test for optional functionality (the test can be unregistered at any time) and I'm not sure we would want to do that.

I could create the test in feed's repository and execute it only in the project's repository, which would make sense, but currently there is no infrastructure to allow it and I don't want to blow the scope on this changee. I suggest waiting for the planned changes to our testing infrastructure (US-4203, US-4207 and US-4210).

What do you think, @Miroslav-Stopka ?

Miroslav-Stopka replied:
Oh, or how it can be complicated to add one test :) You are right, waiting for the planned changes to our testing infrastructure will be the best way. It would be kind of ineffective if we would sacrifice a couple of hours to implement some tests that would then be rewritten.

Miroslav-Stopka commented:
Peter, my review is done. In the beginning, I was a little confused but after that several things have been cleared up. I hope users enjoy the work with the new feeds :)

PetrHeinz commented on packages/framework/src/Controller/Admin/FeedController.php:
```diff
@@ -81,25 +73,25 @@ public function generateAction($feedName, $domainId)
     public function listAction()
     {
-        $feeds = [];
+        $feedsData = [];
```
Because the variable does not contain instances of `FeedInterface`, just arrays for the template...

Miroslav-Stopka replied:
I see

PetrHeinz commented on packages/framework/src/Model/Feed/FeedExport.php:
```diff
@@ -0,0 +1,195 @@
+        return $this->finished;
+    }
+
+    protected function writeBegin(): void
```
It was already in the `FeedXmlWriter` class. Methods `writeHeader` and `writeFooter` might be better - what do you think?

It would require renaming the Twig block in the feed templates. Is the change needed?

Miroslav-Stopka replied:
Leave it as it is now.

PetrHeinz commented on packages/framework/src/Model/Feed/FeedExportDataQueue.php:
```diff
@@ -0,0 +1,60 @@
+
+use Shopsys\FrameworkBundle\Component\Domain\Config\DomainConfig;
+
+class FeedExportDataQueue
```
Well basically it holds the data for creating the FeedExport via `FeedFacade::createFeedExport()`. It holds pairs consisting of an instance of `DomainConfig` and a feed's name as a string.

Other interpretation is that it is a queue used for iterable feed export in `DailyFeedCronModule`.

---

`FeedQueue` would imply it holds instances of `FeedInterface`, which would be misleading.

Can you think of a better name?

Miroslav-Stopka replied:
Propably no. It is kind of confusing - we have queue of pairs. Im used to using queue of objects, not pairs.

PetrHeinz replied:
How about `FeedExportCreationDataQueue` - queue of data to be used to create a `FeedExport`. That should reduce the ambiguity of whether the data are to be used in the feed export, i.e. whether the data are about the products.

How about moving it, along with the two `*CronModule` classes to a new namepsace `Shopsys\FrameworkBundle\Model\Feed\Cron`? This should couple it more strongly with the cron modules, exposing the fact that the queue is made for the `DailyFeedCronModule`.

Would an of these changes help? Or should the behavior of the queue change?

PetrHeinz replied:
I've renamed it to `FeedExportCreationDataQueue` in a1a6d1f5b. Is that OK with you?

PetrHeinz commented on packages/framework/src/Model/Feed/FeedInterface.php:
```diff
@@ -0,0 +1,34 @@
+     * Returns file path to the Twig template of the feed output.
+     *
+     * Path should be relative to the feed bundle, eg. "@FeedBundleName/feed.xml.twig".
+     * Template has to have blocks named "begin", "item" and "end".
```
I just copy-pasted this annotation from `FeedConfigInterface`, but thanks - I was the one who wrote it there :smile:

MattCzerner commented on packages/framework/src/Model/Feed/DailyFeedCronModule.php:
```diff
@@ -0,0 +1,155 @@
+    private $logger;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Feed\FeedExportDataQueue|null
```
Can null even happen? you set it to FeedExportDataQueue in constructor

PetrHeinz replied:
It cannot be null, thanks. Probably a leftover from previous refactoring :)

LukasHeinz commented:
Hi, during wok on this PR, please consider https://github.com/shopsys/shopsys/issues/297

PetrHeinz commented:
Rebased on 7.0.0-alpha3

The first commit 7c24bfb was already reviewed and approved by @Miroslav-Stopka.

Additional changes were done in fixup commits fore easier review:
4bde79e7a - moved ProductUrlsProvider to shopsys/framework as it is required in multiple feeds, renamed to ProductUrlsBatchLoader
2ad7bbb28 - reordered arguments in methods so they are consistent (domain, product) => (product, domain) in newly created or modified classes
201415401 - fixed use after rebase on refactored components

Now working on the other feeds...

PetrHeinz commented:
Zboží feed done in 15ee28caa and CHANGELOG updated along with simple docs in 24be4e1db.

PetrHeinz commented:
Zboží feed done in 15ee28c, CHANGELOG updated along with basic docs in 24be4e1db.

PetrHeinz commented:
PR is ready for review and testing :heavy_check_mark:

vitek-rostislav commented on packages/product-feed-zbozi/src/Model/FeedItem/ZboziFeedItem.php:
```diff
@@ -0,0 +1,248 @@
+    /**
+     * @return string|null
+     */
+    public function getProductno(): ?string
```
should not it be named `getProductNo`?

vitek-rostislav replied:
oh I see it in template...I do not get their naming conventions :smile:

PetrHeinz replied:
Me neither :smile:

PetrHeinz replied:
But we do have `Product::getPartno` and I see that as a similar thing...

vitek-rostislav commented on packages/product-feed-zbozi/src/Model/FeedItem/ZboziFeedItem.php:
```diff
@@ -0,0 +1,248 @@
+    /**
+     * @return string|null
+     */
+    public function getCategoryText(): ?string
```
based on the naming conventions, this should be probably named `getCategorytext` but please do not do that :laughing:

vitek-rostislav commented on packages/product-feed-zbozi/tests/Unit/ZboziFeedItemTest.php:
```diff
@@ -0,0 +1,229 @@
+        self::assertNull($zboziFeedItem->getMaxCpcSearch());
+    }
+
+    public function testZboziFeedItemWithGroupId()
```
same problem as I mentioned for `HeurekaFeedItemTest` - the content of the method does not correspond with it's name

PetrHeinz replied:
Fixed in c30a011b0

vitek-rostislav commented on CHANGELOG.md:
```diff
@@ -24,6 +24,24 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+  - FeedExport is responsible for the actual generation of a file in batches on a specific domain
+  - FeedRenderer is responsible for rendering the feed from Twig template
+  - FeedPathProvider is responsible for providing the correct filepath and url to the specified feed on a domain
+  - ProductUrlsBatchLoader and ProductParametersBatchLoader are responsible for loading product data in batches 
```
please do not forget to fix the original commit message :slightly_smiling_face: 
Speaking of which, you have a typo there ("importand")

PetrHeinz replied:
There is also another typo in commit msg (e944f6e64 ensuring all **fedds** are exportable on all domains via a smoke test). I'll fix it during the rebase - could you help me with that after the review?

vitek-rostislav replied:
of course :slightly_smiling_face:

vitek-rostislav commented on CHANGELOG.md:
```diff
@@ -24,6 +24,24 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+- [#271 - Complete refactoring of feeds functionality](https://github.com/shopsys/shopsys/pull/271)
+  - modules are responsible for querying the data to improve performance
+  - interfaces from package product-feed-interface are not used anymore as they were only important with open-box architecture
+  - only relevant data is fetched from the database, should result in enhanced performance
```
I though that it really helped the performance so why "should"? :slightly_smiling_face:

PetrHeinz replied:
Because I do not have the exact number for all the feeds. I consider it a hunch at this point. But some preliminary test of Google feed shown about eight-fold increase in speed.

vitek-rostislav replied:
ok, fair enough

vitek-rostislav commented on CHANGELOG.md:
```diff
@@ -24,6 +24,24 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
+  - cron modules use the logger for debug information
+  - DailyFeedCronModule is responsible for continuation of the correct feed after waking up
+
+### [shopsys/product-feed-google], [shopsys/product-feed-heureka], [shopsys/product-feed-heureka-delivery], [shopsys/product-feed-zbozi]
```
I am not sure this is a valid notation in our changelog...

PetrHeinz replied:
Me neither. But I'm 100 % positive that I don't want to copy-paste it four times. I can do it if you insist.

vitek-rostislav replied:
let us try to keep it as it is :laughing:

vitek-rostislav commented on docs/introduction/product-feeds.md:
```diff
@@ -0,0 +1,33 @@
+## When are they exported?
+
+Product feeds are usually exported using Cron modules.
+The Cron modules are already implemented and registered, all that's needed it to run the [`cron` phing target](phing-targets.md#cron) every 5 minutes on your server and Shopsys Framework takes care of the rest.
```
~...all that's needed it...~ -> ...all that's needed **is**

PetrHeinz replied:
Fixed in 63c4c6255

PetrHeinz commented on packages/product-feed-heureka/src/HeurekaFeed.php:
```diff
@@ -0,0 +1,51 @@
+     */
+    public function getItems(DomainConfig $domainConfig, ?int $lastSeekId, int $maxResults): iterable
+    {
+        yield from $this->feedItemFacade->getItems($domainConfig, $lastSeekId, $maxResults);
```
I don't see it as that important. The feed would work even when an *return*ed array is used instead of a *yield*ed generator...

vitek-rostislav replied:
I was worried about correct usage of nested `yield`s but I do agree with you that it is not important now.

PetrHeinz commented on packages/product-feed-google/src/GoogleFeedInfo.php:
```diff
@@ -0,0 +1,48 @@
+        return $this->translator->trans(
+            'Google Shopping product feed is not optimized for selling to Australia,
+            Czechia, France, Germany, Italy, Netherlands, Spain, Switzerland, the UK,
+            and the US. It is caused by missing \'shipping\' atribute.'
```
Not my typo, but fixed in 9daefbe13.

PetrHeinz commented on packages/product-feed-google/src/Model/FeedItem/GoogleFeedItem.php:
```diff
@@ -0,0 +1,188 @@
+     */
+    public function getIdentifiers(): array
+    {
+        return array_filter([
```
Yes, the array_filter will filter the filled attributes (i.e. those that `!= false`). Is is unreadable? I've found it elegant...

Would you prefer this?
```php
/**
 * @return string[]
 */
public function getIdentifiers(): array
{
    $identifiers = [];

    if ($this->ean !== null) {
        $identifiers[self::IDENTIFIER_TYPE_EAN] = $this->ean;
    }
    if ($this->partno !== null) {
        $identifiers[self::IDENTIFIER_TYPE_PARTNO] = $this->partno;
    }
    
    return $identifiers;
}
```

Or maybe just emphasize the filter callback (and having it comparing more strictly) like this?
```php
/**
 * @return string[]
 */
public function getIdentifiers(): array
{
    $allIdentifiers = [
        self::IDENTIFIER_TYPE_EAN => $this->ean,
        self::IDENTIFIER_TYPE_PARTNO => $this->partno,
    ];

    return array_filter($allIdentifiers, function ($identifier) {
        return $identifier !== null;
    });
}
```

PetrHeinz commented on packages/product-feed-heureka/src/Model/FeedItem/HeurekaFeedItemFactory.php:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * @var \Shopsys\ProductFeed\HeurekaBundle\Model\FeedItem\HeurekaProductDataBatchLoader
+     */
+    protected $productDataBatchLoader;
```
OK, I'll try to avoid it in the future. For me the `heureka` prefix was not important from the point of view of this class, but I get your point. I'll leave it as-is if you don't object, OK?

PetrHeinz commented on packages/product-feed-heureka/tests/Unit/HeurekaFeedItemTest.php:
```diff
@@ -0,0 +1,222 @@
+        self::assertNull($heurekaFeedItem->getCpc());
+    }
+
+    public function testHeurekaFeedItemWithGroupId()
```
Damn, you're right. Fixed in 51bed1e01.

PetrHeinz commented on packages/product-feed-zbozi/src/Model/FeedItem/ZboziFeedItemFacade.php:
```diff
@@ -0,0 +1,19 @@
+<?php
```
Yes I've noticed this after creating it and didn't find the inner git-strength to fixup a fixup commit.
